src/sparse_blas/backends/mkl_common/mkl_spmv.cxx

```diff
@@ -0,0 +1,279 @@
+    }
+
+    if (A_view.type_view != oneapi::mkl::sparse::matrix_descr::triangular &&
+        A_view.diag_view == oneapi::mkl::diag::unit) {
+        throw mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "`unit` diag_view can only be used with a triangular type_view.");
+    }
+
+    if ((A_view.type_view == oneapi::mkl::sparse::matrix_descr::symmetric ||
+         A_view.type_view == oneapi::mkl::sparse::matrix_descr::hermitian) &&
+        opA == oneapi::mkl::transpose::conjtrans) {
+        throw mkl::unimplemented(
+            "sparse_blas", function_name,
+            "The backend does not support Symmetric or Hermitian matrix with `conjtrans`.");
+    }
+}
+
+void spmv_buffer_size(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                      oneapi::mkl::sparse::matrix_view A_view,
+                      oneapi::mkl::sparse::matrix_handle_t A_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t x_handle, const void *beta,
+                      oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                      oneapi::mkl::sparse::spmv_alg /*alg*/,
+                      oneapi::mkl::sparse::spmv_descr_t spmv_descr, std::size_t &temp_buffer_size) {
+    // TODO: Add support for external workspace once the close-source oneMKL backend supports it.
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    temp_buffer_size = 0;
+    spmv_descr->buffer_size_called = true;
+}
+
+inline void common_spmv_optimize(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                                 oneapi::mkl::sparse::matrix_view A_view,
+                                 oneapi::mkl::sparse::matrix_handle_t A_handle,
+                                 oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+                                 const void *beta,
+                                 oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                                 oneapi::mkl::sparse::spmv_alg alg,
+                                 oneapi::mkl::sparse::spmv_descr_t spmv_descr) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    bool is_beta_host_accessible = detail::is_ptr_accessible_on_host(queue, beta);
+    check_valid_spmv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     is_beta_host_accessible);
+    if (!spmv_descr->buffer_size_called) {
+        throw mkl::uninitialized(
+            "sparse_blas", __func__,
+            "spmv_buffer_size must be called with the same arguments before spmv_optimize.");
+    }
```

Comment:
```suggestion
    if (!spmv_descr->buffer_size_called) {
        throw mkl::uninitialized(
            "sparse_blas", __func__,
            "spmv_buffer_size must be called before spmv_optimize.");
    }
```
The same argument thing applies inside the `CHECK_DESCR_MATCH` macro you have defined. This is a check on whether a previous API has been called or not, regardless of the arguments.

Reply:
This one is still relevant.

Reply:
I thought this message would better describe the intention but anyway this is reworded in https://github.com/oneapi-src/oneMKL/pull/500/commits/0ecb032dc02f141ad57800b591408aa680428cac.
I don't see the issue in the `CHECK_DESCR_MATCH` macro, this will generate messages like "opA argument must match with the previous call to spmm_optimize".

Reply:
Right, I wasn't talking about `CHECK_DESCR_MATCH` being wrong (that one _should_ be about matching previous args), I was only only talking about this highlighted code block where the check _isn't_ about matching with previous args but rather checking if a previous function was called. Latest change was what I wanted. Thanks for fixing it!